[YUNIKORN-3103] Add comprehensive e2e replication test suite for YuniKorn scheduler

rrajesh-cloudera opened the pull request:
### What is this PR for?
- Add 13 comprehensive test cases covering various replication scenarios
- Test ReplicaSets, Deployments, StatefulSets, multi-container pods, node affinity, rolling updates, priority classes, mixed workloads, and stress testing
- Validate YuniKorn scheduler integration and resource allocation
- Include performance optimizations for faster CI/CD execution
- Add DeleteService method to k8s utils for test cleanup
- Optimize timeouts and resource requests for 60% faster execution
- Use parallel pod waiting and single lightweight image for efficiency

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [x] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-3103

### How should this be tested?

### Screenshots (if appropriate)

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

rrajesh-cloudera committed:
Add comprehensive e2e replication test suite for YuniKorn scheduler

- Add 13 comprehensive test cases covering various replication scenarios
- Test ReplicaSets, Deployments, StatefulSets, multi-container pods, node affinity, rolling updates, priority classes, mixed workloads, and stress testing
- Validate YuniKorn scheduler integration and resource allocation
- Include performance optimizations for faster CI/CD execution
- Add DeleteService method to k8s utils for test cleanup
- Optimize timeouts and resource requests for 60% faster execution
- Use parallel pod waiting and single lightweight image for efficiency

rrajesh-cloudera committed:
Fix golangci-lint issues in replication tests

- Replace len() assertions with proper gomega matchers (ginkgolinter)
- Fix formatting issues (gofmt)
- Fix variable shadowing issue (govet)
- Remove unnecessary embedded field access (staticcheck)
- All golangci-lint checks now pass

rrajesh-cloudera commented:
Could you please review the PR ?
@pbacsko @manirajv06

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+
+		ginkgo.By("Waiting for replica pods to be created and running")
+		var podList *v1.PodList
+		gomega.Eventually(func() bool {
+			podList, err = kClient.ListPods(dev, "app=replica-test")
+			if err != nil {
+				return false
+			}
+			return len(podList.Items) == int(replicaCount)
+		}, standardTimeout, pollInterval).Should(gomega.BeTrue(), "Should have 2 replica pods")
```
This code fragment is repeated a lot. We need helper functions inside `KubeCtl`, similar to `WaitForPodBySelectorRunning()`. We can create two methods:

`WaitForPodCountBySelectorRunning()` - wait for the number of pods + all of them should be in running state
`WaitForPodCountBySelector()` - wait for the number of pods (they don't have to be in Running state)

This way, a lot of code duplication inside `gomegy.Eventually()` can be removed and we can just switch to the appropriate `KubeCtl` method (which is what we do anyway).

I haven't tested these but this is a good starting point:
```
func (k *KubeCtl) WaitForPodCountBySelectorRunning(namespace string, selector string, count, timeout int) error {
	var podList *v1.PodList
	var err error

	err = wait.PollUntilContextTimeout(context.TODO(), time.Second, time.Duration(timeout)*time.Second, false,
		func(ctx context.Context) (done bool, err error) {
			podList, err = k.ListPods(namespace, selector)
			if err != nil {
				return false, err
			}
			return len(podList.Items) == count, nil
		})
	if err != nil {
		return err
	}
	
	for _, pod := range podList.Items {
		if err := k.WaitForPodRunning(namespace, pod.Name, time.Duration(timeout)*time.Second); err != nil {
			return err
		}
	}
	return nil
}

func (k *KubeCtl) WaitForPodCountBySelector(namespace string, selector string, count, timeout int) error {
	err := wait.PollUntilContextTimeout(context.TODO(), time.Second, time.Duration(timeout)*time.Second, false,
		func(ctx context.Context) (done bool, err error) {
			podList, err := k.ListPods(namespace, selector)
			if err != nil {
				return false, err
			}
			return len(podList.Items) == count, nil
		})
	if err != nil {
		return err
	}

	return nil
}
```

If you also need the podList, then just return it along with the error.

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+		gomega.Ω(err).NotTo(gomega.HaveOccurred())
+
+		ginkgo.By("Waiting for scaled pods to be created and running")
+		gomega.Eventually(func() bool {
+			podList, err = kClient.ListPods(dev, "app=scaling-test")
+			if err != nil {
+				return false
+			}
+			if len(podList.Items) != int(scaledReplicas) {
+				return false
+			}
+			// Check all pods are running
+			for _, pod := range podList.Items {
+				if pod.Status.Phase != v1.PodRunning {
+					return false
+				}
+			}
+			return true
+		}, standardTimeout, pollInterval).Should(gomega.BeTrue(), "Should have 3 scaled running pods")
```
Replace with the suggested `KubeCtl.WaitFor...()` helper function

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+
+		ginkgo.By("Waiting for all pods to be created and running")
+		var podList *v1.PodList
+		gomega.Eventually(func() bool {
+			podList, err = kClient.ListPods(dev, "app=failure-test")
+			if err != nil {
+				return false
+			}
+			if len(podList.Items) != int(replicaCount) {
+				return false
+			}
+			// Check all pods are running
+			for _, pod := range podList.Items {
+				if pod.Status.Phase != v1.PodRunning {
+					return false
+				}
+			}
+			return true
+		}, standardTimeout, pollInterval).Should(gomega.BeTrue(), "Should have 2 initial running pods")
```
Replace with the suggested `KubeCtl.WaitFor...()` helper function

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+		gomega.Ω(err).NotTo(gomega.HaveOccurred())
+
+		ginkgo.By("Waiting for replacement pod to be created and running")
+		gomega.Eventually(func() bool {
+			podList, err = kClient.ListPods(dev, "app=failure-test")
+			if err != nil {
+				return false
+			}
+			if len(podList.Items) != int(replicaCount) {
+				return false
+			}
+			// Check all pods are running
+			for _, pod := range podList.Items {
+				if pod.Status.Phase != v1.PodRunning {
+					return false
+				}
+			}
+			return true
+		}, standardTimeout, pollInterval).Should(gomega.BeTrue(), "Should have 2 pods after recovery")
```
Replace with the suggested helper function

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+		replicaSetName := "resource-test-replicaset"
+
+		replicaSet := &appsv1.ReplicaSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: replicaSetName,
+				Labels: map[string]string{
+					"app": "resource-test",
+				},
+			},
+			Spec: appsv1.ReplicaSetSpec{
+				Replicas: &replicaCount,
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"app": "resource-test",
+					},
+				},
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{
+							"app": "resource-test",
+						},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:    "resource-container",
+								Image:   testImage,
+								Command: []string{"sleep", "300"},
+								Resources: v1.ResourceRequirements{
+									Requests: v1.ResourceList{
+										v1.ResourceCPU:    resource.MustParse("50m"),  // Reduced from 100m
+										v1.ResourceMemory: resource.MustParse("64Mi"), // Reduced from 128Mi
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
```
Replace with a replicaSet creator helper function.

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+		replicaSetName := "queue-test-replicaset"
+
+		replicaSet := &appsv1.ReplicaSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: replicaSetName,
+				Labels: map[string]string{
+					"app": "queue-test",
+				},
+			},
+			Spec: appsv1.ReplicaSetSpec{
+				Replicas: &replicaCount,
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"app": "queue-test",
+					},
+				},
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{
+							"app": "queue-test",
+						},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:    "queue-container",
+								Image:   testImage,
+								Command: []string{"sleep", "300"},
+								Resources: v1.ResourceRequirements{
+									Requests: v1.ResourceList{
+										v1.ResourceCPU:    resource.MustParse("10m"),
+										v1.ResourceMemory: resource.MustParse("16Mi"),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
```
Replace with a replicaSet creator helper function.

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+						Labels: map[string]string{
+							"app": "multi-container-test",
+						},
+					},
+					Spec: v1.PodSpec{
+						InitContainers: []v1.Container{
+							{
+								Name:    "init-container",
+								Image:   testImage,
+								Command: []string{"sh", "-c", "echo 'Init complete' && sleep 2"}, // Reduced sleep
+							},
+						},
+						Containers: []v1.Container{
+							{
+								Name:    "main-container",
+								Image:   testImage,
+								Command: []string{"sleep", "300"},
+								Resources: v1.ResourceRequirements{
+									Requests: v1.ResourceList{
+										v1.ResourceCPU:    resource.MustParse("25m"),  // Reduced from 50m
+										v1.ResourceMemory: resource.MustParse("32Mi"), // Reduced from 64Mi
+									},
+								},
+							},
+							{
+								Name:    "sidecar-container",
+								Image:   testImage,
+								Command: []string{"sh", "-c", "while true; do echo 'Sidecar running'; sleep 30; done"},
+								Resources: v1.ResourceRequirements{
+									Requests: v1.ResourceList{
+										v1.ResourceCPU:    resource.MustParse("10m"),  // Reduced from 25m
+										v1.ResourceMemory: resource.MustParse("16Mi"), // Reduced from 32Mi
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
```
Replace with a replicaSet creator helper function.

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+
+		ginkgo.By("Waiting for all multi-container pods to be created and running")
+		var podList *v1.PodList
+		gomega.Eventually(func() bool {
+			podList, err = kClient.ListPods(dev, "app=multi-container-test")
+			if err != nil {
+				return false
+			}
+			if len(podList.Items) != int(replicaCount) {
+				return false
+			}
+			// Check all pods are running
+			for _, pod := range podList.Items {
+				if pod.Status.Phase != v1.PodRunning {
+					return false
+				}
+				// Check init container completion
+				if len(pod.Status.InitContainerStatuses) > 0 {
+					initStatus := pod.Status.InitContainerStatuses[0]
+					if initStatus.State.Terminated == nil || initStatus.State.Terminated.ExitCode != 0 {
+						return false
+					}
+				}
+				// Check main containers
+				if len(pod.Status.ContainerStatuses) != 2 {
+					return false
+				}
+				for _, containerStatus := range pod.Status.ContainerStatuses {
+					if !containerStatus.Ready {
+						return false
+					}
+				}
+			}
+			return true
+		}, longTimeout, pollInterval).Should(gomega.BeTrue(), "All multi-container pods should be ready")
```
Ok, this probably needs a different helper function than the previous ones. Regardless, this logic should not be here, but inside `KubeCtl`.

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+						Labels: map[string]string{
+							"app": "node-affinity-test",
+						},
+					},
+					Spec: v1.PodSpec{
+						Affinity: &v1.Affinity{
+							NodeAffinity: &v1.NodeAffinity{
+								PreferredDuringSchedulingIgnoredDuringExecution: []v1.PreferredSchedulingTerm{
+									{
+										Weight: 100,
+										Preference: v1.NodeSelectorTerm{
+											MatchExpressions: []v1.NodeSelectorRequirement{
+												{
+													Key:      "kubernetes.io/hostname",
+													Operator: v1.NodeSelectorOpIn,
+													Values:   []string{targetNode},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+						Containers: []v1.Container{
+							{
+								Name:    "affinity-container",
+								Image:   testImage,
+								Command: []string{"sleep", "300"},
+								Resources: v1.ResourceRequirements{
+									Requests: v1.ResourceList{
+										v1.ResourceCPU:    resource.MustParse("10m"),
+										v1.ResourceMemory: resource.MustParse("16Mi"),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
```
Replace with a replicaSet creator helper function.

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+				Name: replicaSetName,
+				Labels: map[string]string{
+					"app": "stress-test",
+				},
+			},
+			Spec: appsv1.ReplicaSetSpec{
+				Replicas: &replicaCount,
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"app": "stress-test",
+					},
+				},
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{
+							"app": "stress-test",
+						},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:    "stress-container",
+								Image:   testImage,
+								Command: []string{"sleep", "300"},
+								Resources: v1.ResourceRequirements{
+									Requests: v1.ResourceList{
+										v1.ResourceCPU:    resource.MustParse("5m"),  // Reduced from 10m
+										v1.ResourceMemory: resource.MustParse("8Mi"), // Reduced from 16Mi
+									},
+									Limits: v1.ResourceList{
+										v1.ResourceCPU:    resource.MustParse("25m"),  // Reduced from 50m
+										v1.ResourceMemory: resource.MustParse("32Mi"), // Reduced from 64Mi
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
```
Call replicaSet creator function.

pbacsko commented on test/e2e/replication/replication_test.go:
```diff
@@ -0,0 +1,1469 @@
+
+		// Create ReplicaSet (reduced replicas)
+		replicaSet := &appsv1.ReplicaSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "mixed-replicaset",
+				Labels: map[string]string{
+					"app": "mixed-replica",
+				},
+			},
+			Spec: appsv1.ReplicaSetSpec{
+				Replicas: &[]int32{2}[0], // Reduced from 3
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"app": "mixed-replica",
+					},
+				},
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{
+							"app": "mixed-replica",
+						},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:    "replica-container",
+								Image:   testImage,
+								Command: []string{"sleep", "300"},
+								Resources: v1.ResourceRequirements{
+									Requests: v1.ResourceList{
+										v1.ResourceCPU:    resource.MustParse("25m"),  // Reduced from 50m
+										v1.ResourceMemory: resource.MustParse("32Mi"), // Reduced from 64Mi
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
```
Replace with a replicaSet creator function call.

pbacsko reviewed (changes requested):
Hey @rrajesh-cloudera thanks for the PR. My only concern is that the test logic is VERY LONG in the current form and it's difficult to read.
At least let's get rid of the most commonly repeated parts to reduce the size of the tests.

Please look at my suggestions:
1. Create helper methods which return a basic replica set, deployment and stateful set objects with all common fields populated.
2. If you need changes, just call the method ie. `getReplicaSetSpec()` then modify the fields in the testcase code.
3. Replace all `gomega.Eventually()` call with appropriate helper calls from `KubeCtl`. These will be new methods.

rrajesh-cloudera committed:
feat: enhance e2e replication tests with comprehensive scenarios and improved framework

- Fix priority class conflict handling with proper error management for 'already exists' scenarios
- Correct priority value assertions to match actual YuniKorn behavior (1000, 100 values)
- Extend ReplicaSetConfig to support advanced features:
  * Init containers for multi-stage pod initialization
  * Sidecar containers for additional functionality
  * Node affinity rules for targeted scheduling
- Enhance getReplicaSetSpec function to handle complex pod configurations
- Add robust resource cleanup with proper error handling for edge cases
- Implement comprehensive test coverage for 13 replication scenarios:
  * Basic replication setup verification
  * ReplicaSet pod replication and scaling
  * Deployment scaling operations
  * Pod failure recovery mechanisms
  * Resource allocation across replicas
  * Queue assignment consistency
  * Multi-container pod replication
  * Node affinity scheduling
  * Rolling update replication
  * StatefulSet replication with persistent identity
  * Stress testing with multiple replicas
  * Mixed workload replication (concurrent ReplicaSet/Deployment)
  * Priority class-based replication
- Optimize linter performance and resolve all code quality issues
- Add cleanupResource helper function for consistent resource management
- Improve test framework with better error handling and validation

All tests now pass successfully (13 Passed | 0 Failed) with execution time of ~4 minutes.

Closes: Comprehensive e2e replication test suite implementation

pbacsko reviewed (changes requested):
I added some further comments.

I don't think we need all tests here, especially the "nodeAffinity" and "stress" stuff. These can become complicated very quickly.
I also don't see tuch much value in the following: `Verify_Mixed_Workload_Replication`, `Verify_Replication_With_Priority_Classes`.

The checks inside `Verify_Resource_Allocation_Across_Replicas` and `Verify_Queue_Assignment_Consistency` can be merged to 
`Verify_ReplicaSet_Pod_Replication`.

The rest can stay. By making these changes, code will be further reduced, this making it a smaller maintenance burden.

rrajesh-cloudera committed:
refactor: streamline workload e2e tests for better maintainability

- Rename package from 'replication_test' to 'workload_test' for better clarity
- Reduce test count from 13 to 6 focused tests (54% reduction, -449 lines)
- Remove complex tests (nodeAffinity, stress, mixed workload, priority classes)
- Remove basic pod scheduling test (belongs in basic_scheduling suite)
- Merge resource allocation and queue assignment checks into main ReplicaSet test
- Rename test functions for consistency:
  * Verify_Pod_Failure_Recovery → Verify_ReplicaSet_PodRestart
  * Verify_Deployment_Scaling_Operations → Verify_Deployment_Scheduling
  * Verify_ReplicaSet_Pod_Replication → Verify_ReplicaSet_Scheduling
  * Verify_Rolling_Update_Replication → Verify_Deployment_RollingUpdate

- Move workload-specific helper functions from k8s_utils.go to test file:
  * GetBasicDeploymentSpec, GetBasicStatefulSetSpec, GetBasicHeadlessService
  * Reduces k8s_utils.go bloat and keeps helpers close to usage

- Clean up redundant code:
  * Remove unnecessary 'if len() > 0' checks where length already verified
  * Fix indentation and remove excessive blank lines
  * Add robust cleanup error handling

- Update suite configuration:
  * TestReplication → TestWorkload
  * TEST-replication_junit.xml → TEST-workload_junit.xml
  * Describe('Replication') → Describe('Workload Test')

All 6 remaining tests pass successfully (118s execution time):
✓ Verify_ReplicaSet_Scheduling - ReplicaSet with resource & queue verification
✓ Verify_Deployment_Scheduling - Deployment creation and scaling
✓ Verify_ReplicaSet_PodRestart - Pod failure recovery
✓ Verify_Multi_Container_Pod_Replication - Multi-container pods with init containers
✓ Verify_Deployment_RollingUpdate - Rolling update functionality
✓ Verify_StatefulSet_Replication - StatefulSet with persistent identity

Result: Focused, maintainable test suite with comprehensive workload coverage
